cg.go

```diff
@@ -0,0 +1,293 @@
+	// For medium and large-scale problems IterationRestartFactor should be set
+	// to 1 and greater values should be used for low-dimensional problems.
+	// Care must be taken because if the resulting number is 1, the iteration
+	// will be restarted at every iteration and the CG method will reduce to
+	// the steepest descent method.
+	// If IterationRestartFactor is 0, it will be set to 6.
+	IterationRestartFactor float64
+	// The method is restarted whenever the gradients at two consecutive
+	// iterations point in almost opposite directions, i.e., whenever
+	// ∇f(x_k)·∇f(x_{k+1}) <= -AngleRestartThreshold |∇f(x_k)| |∇f(x_{k+1})|.
+	// AngleRestartThreshold should be taken smaller, but close to 1 to reduce
+	// the number of restarts.
+	// If AngleRestartThreshold is 0, it will be set to 0.9.
+	AngleRestartThreshold float64
+
+	linesearch *Linesearch
+
+	restartAfter    int
+	iterFromRestart int
+
+	dirPrev      []float64
+	gradPrev     []float64
+	gradPrevNorm float64
+}
+
+func (cg *CG) Init(loc Location, funcInfo *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	if cg.LinesearchMethod == nil {
+		cg.LinesearchMethod = &Bisection{GradConst: 0.1}
+	}
+	if cg.Variant == nil {
+		cg.Variant = &HestenesStiefel{}
+	}
+	if cg.InitialStep == nil {
+		cg.InitialStep = &FirstOrderStepSize{}
+	}
+
+	if cg.IterationRestartFactor == 0 {
+		cg.IterationRestartFactor = iterationRestartFactor
+	}
+	if cg.AngleRestartThreshold == 0 {
+		cg.AngleRestartThreshold = angleRestartThreshold
+	}
+
+	if cg.linesearch == nil {
+		cg.linesearch = &Linesearch{}
+	}
+	cg.linesearch.Method = cg.LinesearchMethod
+	cg.linesearch.NextDirectioner = cg
+
+	return cg.linesearch.Init(loc, funcInfo, xNext)
+}
+
+func (cg *CG) Iterate(loc Location, xNext []float64) (EvaluationType, IterationType, error) {
+	return cg.linesearch.Iterate(loc, xNext)
+}
+
+func (cg *CG) InitDirection(loc Location, direction []float64) (stepSize float64) {
+	dim := len(loc.X)
+
+	cg.restartAfter = int(math.Ceil(cg.IterationRestartFactor * float64(dim)))
+	cg.iterFromRestart = 0
+
+	// The initial direction is always the steepest descent direction.
+	copy(direction, loc.Gradient)
+	floats.Scale(-1, direction)
+
+	cg.dirPrev = resize(cg.dirPrev, dim)
+	copy(cg.dirPrev, direction)
+	cg.gradPrev = resize(cg.gradPrev, dim)
+	copy(cg.gradPrev, loc.Gradient)
+	cg.gradPrevNorm = floats.Norm(loc.Gradient, 2)
+
+	cg.Variant.Init(loc)
+	return cg.InitialStep.Init(loc, direction)
+}
+
+func (cg *CG) NextDirection(loc Location, direction []float64) (stepSize float64) {
+	copy(direction, loc.Gradient)
+	floats.Scale(-1, direction)
+
+	cg.iterFromRestart++
+	var restart bool
+	// Check if the method should be restarted because too many iterations have
+	// been taken without a restart.
+	if cg.iterFromRestart%cg.restartAfter == 0 {
+		restart = true
+	}
+	// Check if the method should be restarted because the last two gradients
+	// point in almost opposite directions.
```

Comment:
Yea, you're right, I got confused with the negative sign. My fault.

Would it make more sense to have the restart happen if the angle is less than AngleThreshhold and remove the negative sign? In this case, AngleThreshhold would default to -0.9. This way it aligns with the Cos formula, 1 and they're perfectly aligned, 0 they're othogonal, -1 they're exactly opposite.